src/HealthGPS/risk_factor_adjustable_model.cpp

```diff
@@ -0,0 +1,170 @@
+#include "HealthGPS.Core/exception.h"
+#include "HealthGPS.Core/thread_util.h"
+
+#include "risk_factor_adjustable_model.h"
+#include "sync_message.h"
+
+namespace { // anonymous namespace
+
+/// @brief Defines the risk factors adjustment synchronisation message
+using RiskFactorAdjustmentMessage = hgps::SyncDataMessage<hgps::RiskFactorSexAgeTable>;
+
+/// @brief Defines the first statistical moment type
+struct FirstMoment {
+  public:
+    void append(double value) noexcept {
+        count_++;
+        sum_ += value;
+    }
+
+    double mean() const noexcept { return (count_ > 0) ? (sum_ / count_) : 0.0; }
+
+  private:
+    int count_{};
+    double sum_{};
+};
+
+} // anonymous namespace
+
+namespace hgps {
+
+RiskFactorAdjustableModel::RiskFactorAdjustableModel(
+    const RiskFactorSexAgeTable &risk_factor_expected)
+    : risk_factor_expected_{risk_factor_expected} {
+
+    if (risk_factor_expected_.empty()) {
+        throw core::HgpsException("Risk factor expected value mapping is empty");
+    }
+}
+
+const RiskFactorSexAgeTable &RiskFactorAdjustableModel::get_risk_factor_expected() const noexcept {
+    return risk_factor_expected_;
+}
+
+void RiskFactorAdjustableModel::adjust_risk_factors(
+    RuntimeContext &context, const std::vector<core::Identifier> &keys) const {
+    RiskFactorSexAgeTable adjustments;
+
+    // Baseline scenatio: compute adjustments.
+    if (context.scenario().type() == ScenarioType::baseline) {
+        adjustments = calculate_adjustments(context, keys);
+    }
+
+    // Intervention scenario: recieve adjustments from baseline scenario.
+    else {
+        auto message = context.scenario().channel().try_receive(context.sync_timeout_millis());
+        if (!message.has_value()) {
+            throw core::HgpsException(
+                "Simulation out of sync, receive baseline adjustments message has timed out");
+        }
+
+        auto &basePtr = message.value();
+        auto *messagePrt = dynamic_cast<RiskFactorAdjustmentMessage *>(basePtr.get());
```

Comment:
I'm quite lost with this line. Could you clarify what's going on here?

Reply:
`dynamic_cast` lets you cast from a parent class to a child class, which you otherwise wouldn't be able to do. If it *isn't* an instance of your child class, it returns `nullptr`.